driver-core/src/main/com/mongodb/internal/connection/MongoCredentialWithCache.java

```diff
@@ -63,28 +63,23 @@ public void putInCache(final Object key, final Object value) {
         cache.set(key, value);
     }
 
-    public Lock getLock() {
-        return cache.lock;
+    public <V> V withLock(final Supplier<V> k) {
+        return Locks.withLock(cache.lock, k);
     }
 
     static class Cache {
         private final ReentrantLock lock = new ReentrantLock();
-        private Object cacheKey;
-        private Object cacheValue;
+        private final ConcurrentHashMap<Object, Object> cache = new ConcurrentHashMap<>();
```

Comment:
Does it need to be concurrent since access is protected by a lock anyway?

Reply:
Alternatively consider removing the lock.

Reply:
Reverted to old pair and documented.